ggshield/cmd/sca/scan.py

```diff
@@ -57,14 +51,119 @@ def scan_diff_cmd(
+    update_context(ctx)
+
+    result = sca_scan_all(ctx, directory)
+
+    # TODO, should be handled by an output_handler in the future
+    return result
+
+
+def update_context(ctx: click.Context) -> None:
+    """
+    Basic implementation, should be populated with required parameters later.
+    """
+    config: Config = ctx.obj["config"]
+    ctx.obj["client"] = create_client_from_config(config)
+
+    # Empty for now
+    ctx.obj["exclusion_regexes"] = set()
+
+
+def sca_scan_all(ctx: click.Context, directory: Path) -> Optional[SCAScanAllOutput]:
+    """
+    Scan an entire directory for SCA Vulnerabilities.
+
+    - List SCA related files with a first call to SCA compute files API
+    - Create a tar archive with the required files contents
+    - Launches the scan with a call to SCA public API
+    """
+    client = SCAClient(ctx.obj["client"])
+
+    sca_filepaths = get_sca_scan_all_filepaths(
+        directory=directory,
+        exclusion_regexes=ctx.obj["exclusion_regexes"],
+        verbose=ctx.obj["config"].verbose,
+        client=client,
+    )
+
+    # No file to scan
+    if sca_filepaths is None or len(sca_filepaths) == 0:
+        display_info("No file to scan.")
+        return
+
+    # empty for now
+    scan_parameters = SCAScanParameters()
+
+    tar = _create_tar(directory, sca_filepaths)
+
+    # Call to full scan API and get results
+    scan_result = client.sca_scan_directory(
+        tar,
+        scan_parameters,
+        ScanContext(
+            command_path=ctx.command_path,
+            scan_mode=ScanMode.SCA_DIRECTORY,
+        ).get_http_headers(),
+    )
+
+    if not isinstance(scan_result, SCAScanAllOutput):
+        if scan_result.status_code == 401:
+            raise APIKeyCheckError(client.base_uri, "Invalid API key.")
+        display_error("Error scanning.")
+        display_error(str(scan_result))
+        return
+
+    return scan_result
+
+
+def get_sca_scan_all_filepaths(
+    directory: Path,
+    exclusion_regexes: Set[re.Pattern],
+    verbose: bool,
+    client: SCAClient,
+) -> Optional[List[str]]:
+    """
+    Retrieve SCA related files of a directory.
+    First get all filenames that are not in blacklisted directories, then calls SCA compute files
+    API to filter SCA related files.
+    """
+    all_filepaths = get_all_files_from_sca_paths(
+        path=directory,
+        exclusion_regexes=exclusion_regexes,
+        verbose=verbose,
+        # If the repository is a git repository, ignore untracked files
+        ignore_git=True,
+    )
+
+    # API Call to filter SCA files
+    response = client.compute_sca_files(touched_files=all_filepaths)
+
+    if not isinstance(response, ComputeSCAFilesResult):
+        if response.status_code == 401:
+            raise APIKeyCheckError(client.base_uri, "Invalid API key.")
+        display_error("Error while filtering SCA related files.")
+        display_error(str(response))
+        return None
```

Comment:
Indeed. I raise an error if `compute_sca_files` fails, and return an empty `SCAScanAllOutput` in `sca_scan_all` in case of an empty result from `get_sca_scan_all_filepaths`. WDYT ?

Reply:
Sorry I had not seen your comment. 
> return an empty SCAScanAllOutput in sca_scan_all in case of an empty result from get_sca_scan_all_filepaths

I do not really see the point. If an error happens in `get_sca_scan_all_filepaths`, there is no point returning an SCAScanAllOutput - something went wrong, the only thing we can do is output meaningful error message and exit code.

Reply:
No, if we return an empty list, no error occured. If we return an empty list, there is just no file to scan, so in this case we do not return `None` anymore, we return an empty `SCAScanAllOutput` because it is basically the expected result in this case.

In case of an error, we raise an `UnexpectedError`, and we do not reach the line where we return an empty `SCAScanAllOutput`. (In the previous proposition in case of error I did return `None`, now I just raise).

Reply:
The misunderstanding maybe came from the term `empty result` instead of `empty list`.

Reply:
Oh ok, yes I get it, but since we'll probably need a try/catch at the upper level, it's not worth handling errors in `scan_all` with an optional return type IMO. Let's wait, but we can probably handle the ReadTimeout there as well, instead of inside the client.